src/test/run-pass/drop/dynamic-drop-async.rs

```diff
@@ -0,0 +1,315 @@
+// run-pass
```

Comment:
This test is pretty complicated (even more so than the `dynamic-drop.rs` one) and currently requires a lot of time to understand what is going on... In particular, the test harness that is setup here needs some reading back and forth but the overall role of the test could also be made more clear.

To that end, could you please add a comment to the top of the file explaining its role as well as comments throughout `run_test`, `Allocator`, and `Ptr`?

If you don't mind, some light weight comments on the `async fn`s could also be useful for future readers.